File: packages/ios-app/SliccFollower/Models/TrayTypes.swift
import Foundation

// MARK: - TraySessionDescription

struct TraySessionDescription: Codable {
    let type: SDPType
    let sdp: String

    enum SDPType: String, Codable {
        case offer
        case answer
    }
}

// MARK: - TrayIceCandidate

struct TrayIceCandidate: Codable {
    let candidate: String
    let sdpMid: String?
    let sdpMLineIndex: Int?
    let usernameFragment: String?
}

// MARK: - TrayBootstrapState

enum TrayBootstrapState: String, Codable {
    case pending
    case offered
    case connected
    case failed
}

// MARK: - TrayBootstrapFailure

struct TrayBootstrapFailure: Codable {
    let code: String
    let message: String
    let retryable: Bool
    let retryAfterMs: Int?
    let failedAt: String
}

// MARK: - TrayBootstrapStatus

struct TrayBootstrapStatus: Codable {
    let controllerId: String
    let bootstrapId: String
    let attempt: Int
    let state: TrayBootstrapState
    let expiresAt: String
    let cursor: Int
    let maxRetries: Int
    let retriesRemaining: Int
    let retryAfterMs: Int?
    let failure: TrayBootstrapFailure?
}

// MARK: - TurnIceServer

struct TurnIceServer: Codable {
    let urls: [String]
    let username: String
    let credential: String
}

// MARK: - TrayLeaderSummary

struct TrayLeaderSummary: Codable {
    let controllerId: String
    let connected: Bool
    let reconnectDeadline: String?
}

// MARK: - TrayBootstrapEvent

enum TrayBootstrapEvent: Codable {
    case offer(sequence: Int, sentAt: String, offer: TraySessionDescription)
    case iceCandidate(sequence: Int, sentAt: String, candidate: TrayIceCandidate)
    case failed(sequence: Int, sentAt: String, failure: TrayBootstrapFailure)

    private enum CodingKeys: String, CodingKey {
        case type, sequence, sentAt, offer, candidate, failure
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)
        let sequence = try container.decode(Int.self, forKey: .sequence)
        let sentAt = try container.decode(String.self, forKey: .sentAt)
        switch type {
        case "bootstrap.offer":
            let offer = try container.decode(TraySessionDescription.self, forKey: .offer)
            self = .offer(sequence: sequence, sentAt: sentAt, offer: offer)
        case "bootstrap.ice_candidate":
            let cand = try container.decode(TrayIceCandidate.self, forKey: .candidate)
            self = .iceCandidate(sequence: sequence, sentAt: sentAt, candidate: cand)
        case "bootstrap.failed":
            let fail = try container.decode(TrayBootstrapFailure.self, forKey: .failure)
            self = .failed(sequence: sequence, sentAt: sentAt, failure: fail)
        default:
            throw DecodingError.dataCorrupted(
                .init(codingPath: decoder.codingPath,
                      debugDescription: "Unknown TrayBootstrapEvent type: \(type)"))
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .offer(seq, sent, offer):
            try container.encode("bootstrap.offer", forKey: .type)
            try container.encode(seq, forKey: .sequence)
            try container.encode(sent, forKey: .sentAt)
            try container.encode(offer, forKey: .offer)
        case let .iceCandidate(seq, sent, cand):
            try container.encode("bootstrap.ice_candidate", forKey: .type)
            try container.encode(seq, forKey: .sequence)
            try container.encode(sent, forKey: .sentAt)
            try container.encode(cand, forKey: .candidate)
        case let .failed(seq, sent, fail):
            try container.encode("bootstrap.failed", forKey: .type)
            try container.encode(seq, forKey: .sequence)
            try container.encode(sent, forKey: .sentAt)
            try container.encode(fail, forKey: .failure)
        }
    }
}

// MARK: - FollowerAttachResult

enum FollowerAttachResult: Codable {
    case wait(code: String, retryAfterMs: Int)
    case signal(code: String, bootstrap: TrayBootstrapStatus)
    case fail(code: String, error: String)

    private enum CodingKeys: String, CodingKey {
        case action, code, retryAfterMs, bootstrap, error
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let action = try container.decode(String.self, forKey: .action)
        switch action {
        case "wait":
            self = .wait(
                code: try container.decode(String.self, forKey: .code),
                retryAfterMs: try container.decode(Int.self, forKey: .retryAfterMs))
        case "signal":
            self = .signal(
                code: try container.decode(String.self, forKey: .code),
                bootstrap: try container.decode(TrayBootstrapStatus.self, forKey: .bootstrap))
        case "fail":
            self = .fail(
                code: try container.decode(String.self, forKey: .code),
                error: try container.decode(String.self, forKey: .error))
        default:
            throw DecodingError.dataCorrupted(
                .init(codingPath: decoder.codingPath,
                      debugDescription: "Unknown action: \(action)"))
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .wait(code, retryAfterMs):
            try container.encode("wait", forKey: .action)
            try container.encode(code, forKey: .code)
            try container.encode(retryAfterMs, forKey: .retryAfterMs)
        case let .signal(code, bootstrap):
            try container.encode("signal", forKey: .action)
            try container.encode(code, forKey: .code)
            try container.encode(bootstrap, forKey: .bootstrap)
        case let .fail(code, error):
            try container.encode("fail", forKey: .action)
            try container.encode(code, forKey: .code)
            try container.encode(error, forKey: .error)
        }
    }
}

// MARK: - FollowerAttachResponse

struct FollowerAttachResponse: Codable {
    let trayId: String
    let controllerId: String
    let role: String
    let leader: TrayLeaderSummary?
    let participantCount: Int
    let result: FollowerAttachResult
    let iceServers: [TurnIceServer]?
}

// MARK: - FollowerBootstrapResponse

struct FollowerBootstrapResponse: Codable {
    let trayId: String
    let controllerId: String
    let role: String
    let leader: TrayLeaderSummary?
    let participantCount: Int
    let bootstrap: TrayBootstrapStatus
    let events: [TrayBootstrapEvent]
    let iceServers: [TurnIceServer]?
}

File: packages/webapp/src/scoops/tray-types.ts
export type TrayBootstrapState = 'pending' | 'offered' | 'connected' | 'failed';

export interface TraySessionDescription {
  type: 'offer' | 'answer';
  sdp: string;
}

export interface TrayIceCandidate {
  candidate: string;
  sdpMid?: string | null;
  sdpMLineIndex?: number | null;
  usernameFragment?: string | null;
}

export interface TrayBootstrapFailure {
  code: string;
  message: string;
  retryable: boolean;
  retryAfterMs: number | null;
  failedAt: string;
}

export type TrayBootstrapEvent =
  | {
      sequence: number;
      sentAt: string;
      type: 'bootstrap.offer';
      offer: TraySessionDescription;
    }
  | {
      sequence: number;
      sentAt: string;
      type: 'bootstrap.ice_candidate';
      candidate: TrayIceCandidate;
    }
  | {
      sequence: number;
      sentAt: string;
      type: 'bootstrap.failed';
      failure: TrayBootstrapFailure;
    };

export interface TrayBootstrapStatus {
  controllerId: string;
  bootstrapId: string;
  attempt: number;
  state: TrayBootstrapState;
  expiresAt: string;
  cursor: number;
  maxRetries: number;
  retriesRemaining: number;
  retryAfterMs: number | null;
  failure: TrayBootstrapFailure | null;
}

export interface TurnIceServer {
  urls: string[];
  username: string;
  credential: string;
}

export interface FollowerJoinRequestedMessage {
  type: 'follower.join_requested';
  trayId: string;
  controllerId: string;
  runtime?: string;
  bootstrapId: string;
  attempt: number;
  expiresAt: string;
  iceServers?: TurnIceServer[];
}

export interface BootstrapAnswerMessage {
  type: 'bootstrap.answer';
  trayId: string;
  controllerId: string;
  bootstrapId: string;
  answer: TraySessionDescription;
}

export interface BootstrapIceCandidateMessage {
  type: 'bootstrap.ice_candidate';
  trayId: string;
  controllerId: string;
  bootstrapId: string;
  candidate: TrayIceCandidate;
}

export interface WebhookEventMessage {
  type: 'webhook.event';
  webhookId: string;
  headers: Record<string, string>;
  body: unknown;
  timestamp: string;
}

export type WorkerToLeaderControlMessage =
  | {
      type: 'leader.connected';
      trayId: string;
      controllerId: string;
    }
  | {
      type: 'pong';
      trayId: string;
    }
  | FollowerJoinRequestedMessage
  | BootstrapAnswerMessage
  | BootstrapIceCandidateMessage
  | WebhookEventMessage
  | WorkerPreviewRequest
  | WorkerPreviewRevoked;

export interface WorkerPreviewRequest {
  type: 'preview.request';
  reqId: string;
  servedRoot: string;
  vfsPath: string;
  asText: boolean;
}

export interface WorkerPreviewRevoked {
  type: 'preview.revoked';
  previewToken: string;
}

export interface LeaderBootstrapOfferMessage {
  type: 'bootstrap.offer';
  controllerId: string;
  bootstrapId: string;
  offer: TraySessionDescription;
}

export interface LeaderBootstrapIceCandidateMessage {
  type: 'bootstrap.ice_candidate';
  controllerId: string;
  bootstrapId: string;
  candidate: TrayIceCandidate;
}

export interface LeaderBootstrapFailedMessage {
  type: 'bootstrap.failed';
  controllerId: string;
  bootstrapId: string;
  code: string;
  message: string;
  retryable?: boolean;
  retryAfterMs?: number | null;
}

export interface LeaderPreviewResponseOk {
  type: 'preview.response';
  reqId: string;
  ok: true;
  mime: string;
  chunkIndex: number;
  totalChunks: number;
  content: string;
  encoding: 'utf-8' | 'base64';
}

export interface LeaderPreviewResponseError {
  type: 'preview.response';
  reqId: string;
  ok: false;
  status: 404 | 403 | 500;
  reason?: string;
}

export type LeaderPreviewPurge = {
  type: 'preview.purge';
  previewToken: string;
};

export type LeaderToWorkerControlMessage =
  | { type: 'ping' }
  | LeaderBootstrapOfferMessage
  | LeaderBootstrapIceCandidateMessage
  | LeaderBootstrapFailedMessage
  | LeaderPreviewResponseOk
  | LeaderPreviewResponseError
  | LeaderPreviewPurge;

export interface TrayLeaderSummary {
  controllerId: string;
  connected: boolean;
  reconnectDeadline: string | null;
}

export interface FollowerJoinRequest {
  controllerId?: string;
  runtime?: string;
}

export type FollowerAttachResult =
  | {
      action: 'wait';
      code: 'LEADER_NOT_ELECTED' | 'LEADER_NOT_CONNECTED';
      retryAfterMs: number;
    }
  | {
      action: 'signal';
      code: 'LEADER_CONNECTED';
      bootstrap: TrayBootstrapStatus;
    }
  | {
      action: 'fail';
      code: 'INVALID_JOIN_CAPABILITY' | 'TRAY_EXPIRED';
      error: string;
    };

export interface FollowerAttachResponse {
  trayId: string;
  controllerId: string;
  role: 'follower';
  leader: TrayLeaderSummary | null;
  participantCount: number;
  result: FollowerAttachResult;
  iceServers?: TurnIceServer[];
}

export interface FollowerBootstrapResponse {
  trayId: string;
  controllerId: string;
  role: 'follower';
  leader: TrayLeaderSummary | null;
  participantCount: number;
  bootstrap: TrayBootstrapStatus;
  events: TrayBootstrapEvent[];
  iceServers?: TurnIceServer[];
}
